File: Day35/Part3/shortest_subarray_at_least_k.cpp
#include <bits/stdc++.h>
using namespace std;

int shortest_subarray(vector<int> nums, long long K){
    int n = (int)nums.size();
    vector<long long> P(n+1,0);
    for(int i=0;i<n;++i) P[i+1]=P[i]+nums[i];
    deque<int> dq;
    int ans = n+1;
    for(int i=0;i<=n;++i){
        while(!dq.empty() && P[i]-P[dq.front()]>=K){
            ans = min(ans, i - dq.front());
            dq.pop_front();
        }
        while(!dq.empty() && P[i] <= P[dq.back()]) dq.pop_back();
        dq.push_back(i);
    }
    return ans<=n?ans:-1;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << shortest_subarray(vector<int>{2,-1,2}, 3) << "\n"; // Expected 3
    return 0;
}

File: Day35/Part3/shortest_subarray_at_least_k.py
from collections import deque
from typing import List

def shortest_subarray(nums: List[int], K: int) -> int:
    n = len(nums)
    P = [0] * (n + 1)
    for i in range(n):
        P[i+1] = P[i] + nums[i]
    ans = n + 1
    dq = deque()
    for i in range(n + 1):
        while dq and P[i] - P[dq[0]] >= K:
            ans = min(ans, i - dq.popleft())
        while dq and P[i] <= P[dq[-1]]:
            dq.pop()
        dq.append(i)
    return ans if ans <= n else -1

if __name__ == "__main__":
    print(shortest_subarray([2,-1,2], 3))  # Expected 3
